drivers/timer/wch_systick_ch32v00x.c

```diff
@@ -27,20 +26,36 @@
 
 #define SYSTICK ((SysTick_Type *)(DT_INST_REG_ADDR(0)))
 
+static struct k_spinlock lock;
 static volatile uint32_t ch32v00x_systick_count;
 
 static void ch32v00x_systick_irq(const void *unused)
 {
 	ARG_UNUSED(unused);
 
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	SYSTICK->CTLR &= ~(STK_STE);
 	SYSTICK->SR = 0;
+	SYSTICK->CNT = 0;
+	SYSTICK->CTLR |= STK_STE;
+
 	ch32v00x_systick_count += CYCLES_PER_TICK; /* Track cycles. */
+
+	k_spin_unlock(&lock, key);
 	sys_clock_announce(1);                     /* Poke the scheduler. */
 }
 
 uint32_t sys_clock_cycle_get_32(void)
 {
-	return ch32v00x_systick_count + SYSTICK->CNT;
+	uint32_t ret;
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+
+	ret = ch32v00x_systick_count + SYSTICK->CNT;
```

Comment:
Note that time can still jump backwards with this code, as CNT may match, get reset to zero, and ch32v00x_systick_count hasn't been updated.

Good news is the problem already exists :)

If you want, you could change the counter to not reset on match, but instead increase STK_CMPR by CYCLES_PER_TICK on each interrupt.